[VF][Polaris] Update IndexTable Row to support indentation for nested child rows

oksanashopify opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

Resolves first acceptance criteria from https://github.com/Shopify/temp-project-mover-syfiawoo-20250331132715/issues/401 - "IndexTable rows with a rowType of child are indented to represent the nested structure".

Depends on https://github.com/Shopify/polaris/pull/9445
Issue with  IndexTable with Sub Headers does not have proper background-color for unselectable table first-child cell will be fixed in separate [PR.](https://github.com/Shopify/polaris/pull/10960)
 

### WHAT is this pull request doing?

1. Adds New type `child` for a `rowType` to represent the nested structure for child rows
2. Adds Styles (indentation) for child rows
3. Adds stories to tophat changes on different tables

**These changes will allow us to integrate design for new Variants card and also will help other Polaris consumers visually distinct child rows under the sub headers**
 
New Variants card Design: [Figma ](https://www.figma.com/file/PKcKNWig5d7yMoq4iR6nlH/%F0%9F%92%AF-250-Variants-UX?node-id=2835%3A82451&mode=dev)

Desktop:
<img width="284" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/66b01e51-d3a4-47b9-9771-cdf1b57f9ab4">

Mobile:
<img width="323" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/60f37202-bdac-4e44-ae18-52f048bf58ab">

How current snapshot looks on PDP new Variants Card:
<img width="668" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/7f70f98e-7501-44c0-9c82-03ecc5708e76">
[Spin with latest snapshot](https://admin.web.oa-test-polaris-nested-rows.oksana-azarova.us.spin.dev/store/shop1/products/1)
[Draft PR with latest snapshot and child rows](https://github.com/Shopify/web/pull/109999)

Updated website examples and documentation:
<img width="1209" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/cf712738-ad55-4ec5-b895-e2ecab045250">

<img width="908" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/4a415f95-6135-4b06-8082-f6ab6a57e0b4">



<!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩
I would recommend start reviewing from the last two stories, to understand better our initial idea :)

**_Special test case to test UI changes for new Variants Card design_**
1.  Open Storybook:
<img width="785" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/0e3ae182-7566-49b8-a0f4-fb1998afc248">

4. Scroll to last two stories 
<img width="398" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/060144ea-30e6-414a-8d84-d5d236bacfd1">

5. See how Image and Name located in one cell. This is how currently new Variants card is built
6. Test Desktop UI with `With Nested Rows With Thumbnails One Cell Selectable` story
Desktop: [Figma](https://www.figma.com/file/PKcKNWig5d7yMoq4iR6nlH/%F0%9F%92%AF-250-Variants-UX?node-id=4827%3A146409&mode=dev)

7. Test mobile version with `With Nested Rows With Thumbnails One Cell Selectable`  story. In Admin on PDP for single location store we switch to unselectable table with thumbnails for mobile version on consumer level. So that's why I created separate story for that UI
Mobile: [Figma](https://www.figma.com/file/PKcKNWig5d7yMoq4iR6nlH/%F0%9F%92%AF-250-Variants-UX?node-id=4827%3A149172&mode=dev)

8. See how current changes work on web:
[Spin with latest snapshot](https://admin.web.oa-test-polaris-nested-rows.oksana-azarova.us.spin.dev/store/shop1/products/1)
https://github.com/Shopify/web/pull/109999


**General test case:**
1. Open Storybook:
<img width="785" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/6b269773-3c58-471a-8c4b-2a2c53a809d4">

2. Go through each story related to nested child rows:
<img width="389" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/108d174d-c164-4774-bae2-5a53bcb8af3a">

3. Check nested row in the following stories (select, hover rows):

- With Row Status
- With Zebra Striping



5. Test different responsive dimensions.


🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

 
### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

oksanashopify committed:
Added child row type, styles and stories for it

added styles to align checkbox go right below second parent cell

added showBorderTopForChildRow prop to the Row and styles for it

removed show dividerr property and added new story with sub rows

declared variables and calculated cells offset for child rows

removed hardcoded prop value

added description for child row and fixed responsiveness

Revert "added description for child row and fixed responsiveness"

This reverts commit 486431b719b0c1cc01f419a4a95ab246a8afc8a1.

description for child row and fixed responsiveness

added test

removed extra condition operator

added se23 styles

improved indentation calculation for child rows

code clean up

added story with thumbnail and styles for responsiveness

added styles for Table-sticky child row first cell

added story wih non-selectable rows

added more stories

fixed headers in stories

updated thumbnail sizes in stories

added back styles and fixed stoies

fixed styled for next after first cell

Added box padding for cell with thumbnails

cleaned up styes for unselectable rows, aligned thumbnails in one cell solution

refactored stories to use products arrays

removed unused imports

fixed headings for last two stories

fixed styles for next after first cell

added example with longer product attributes

added table with nested rows to examples

fixed story for accessibility

added missing story aftre merging conflicts

remuved subheaders for new Variants example stories

fixed transparent gap next to checkbox when table becomes scrollable

code cleanup after reviews

CI Accessibility test fix

added changeset and reoved scopes for accessibility CI test

changed subHeaderRange to existin property selectionRange

rebased from main

Revert "added changeset and reoved scopes for accessibility CI test"

This reverts commit fe7880f1143466dbacc54b4d70f16d604b4480ba.

removed unused scopes

added changeset

removed extra headings

removed extra headings

removed extra syntax rules

removed extra scopes

removed extra syntax rules

removed extra syntax rules

removed extra syntax rules

fixes for stylelint

fixes for stylelint

fixes for stylelint

fixed offset for selectable tables

fixed styles for unselectable nested rows

removed extra headers

fixed stylelint rules

tatianau commented on polaris-react/src/components/IndexTable/components/Row/Row.tsx:
```diff
@@ -115,7 +115,8 @@ export const Row = memo(function Row({
     styles.TableRow,
-    rowType === 'subheader' && styles['TableRow-subheader'],
+    (rowType === 'subheader' && styles['TableRow-subheader']) ||
+      (rowType === 'child' && styles['TableRow-child']),
```
We shouldn't need this as the modifier classes exist on their own depending on a `rowType`

```suggestion
    rowType === 'subheader' && styles['TableRow-subheader'])
    rowType === 'child' && styles['TableRow-child'])
```

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,27 @@ $loading-panel-height: 53px;
+      left: calc(
+        var(--pc-checkbox-offset) + var(--pc-index-table-checkbox-child-offset)
+      );
+    }
```
```suggestion

    .TableCell-first {
      left: var(--pc-index-table-checkbox-child-offset);
    }
    
    .TableCell-first + .TableCell {
      left: calc(
        var(--pc-checkbox-offset) + var(--pc-index-table-checkbox-child-offset)
      );
    }
```

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
+        var(--pc-index-table-checkbox-width)
+    );
+
+    #{$se23} & {
```
Styles for polarisSummerEditions2023

oksanashopify replied:
Was removed with new Polaris version

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
   }
 
+  &.TableRow-child {
+    cursor: default;
```
Could you provide a story for tophatting nested rows that are non-selectable `selectable={false}` to see how this change scales and whether it needs some other special treatment for when there are no checkboxes but the row are still nested?

oksanashopify replied:
Added additional stories for unselectable tables:
![image](https://github.com/Shopify/polaris/assets/104942025/a9ad3a09-32b6-471b-9527-dc9feff7a05a)


+ Additional story  **With Nested Rows Last Column Sticky**, as it was recommended  [here in Polaris channel](https://shopify.slack.com/archives/C4Y8N30KD/p1695865362886549?thread_ts=1695830302.207619&cid=C4Y8N30KD)

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
     }
   }
 
+  &.TableRow-child {
```
In our mock, the indentation on mobile is smaller than in the storybook, can we verify the spacing for smaller screens with @supervee and @Shopify/polaris-ux ?

<table>
<tr><td>Figma</td><td>Actual</td></tr>
<tr><td><img width="531" alt="Screenshot 2023-09-25 at 3 44 25 PM" src="https://github.com/Shopify/polaris/assets/15054990/80fbbb47-1c8c-4d38-9009-fb101cf666ae">
</td><td><img width="332" alt="Screenshot 2023-09-25 at 3 44 14 PM" src="https://github.com/Shopify/polaris/assets/15054990/b8c31922-a353-4845-8e7c-433056d51c35">
</td></tr>
</table>

supervee replied:
The idea was that on desktop, you'd align the left of the checkbox in the child rows with the title of the parent. 
On mobile, you'd align the right side of the child thumbnail with the right side of the parent image, so that all the titles could be aligned

oksanashopify replied:
We didn't get direct suggestions how rows should look like for other test cases, like unselectable tables, tables without thumbnails, so I added as mush stories as possible so we can test together and see if it works for us

[Thread in Polaris channel](https://shopify.slack.com/archives/C4Y8N30KD/p1695830302207619)

oksanashopify replied:
Currently I can see the following issue with having indentation for mobile devises. As we use positioning `left ` for the first column with checkbox, when table becomes scrollable, next table cell content flows behind that checkbox from the left. Video:

https://github.com/Shopify/polaris/assets/104942025/1a27140b-32ae-4c6d-a08c-8d3fb65d5fc8

Does it make sense to remove indentation for mobile version (`breakpoints-xs (P)`) at all? But it would not resolve an issue when it can become scrollable on wider devices.

I would appreciate any suggestions at that matter.

oksanashopify replied:
Thanks to @tatianau 's suggestion for using `::before` pseudo-element I managed to fix UI issue.
Here is how it looks now:

https://github.com/Shopify/polaris/assets/104942025/946733d4-a3df-46ab-a199-ad34edfdb394

It's probably not ideal look, but as mobile table view is something that we control on consumer level (A[n example with Product Index](https://github.com/Shopify/web/blob/81dd4fc89489fcb6597f040d59490f017216a6d1/app/sections/Products/ProductIndex/components/ProductIndexTable/ProductIndexTable.tsx#L103))  we can swap to different table layout on some specific breakpoint. But at least this fix ensure table would not look broken.

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
   }
 
+  &.TableRow-child {
+    cursor: default;
```
Why do we need to set a mouse cursor here? Shouldn't it inherit what it already has for a selectable row (`pointer`) since the entire row is selectable?

```suggestion
```

oksanashopify replied:
removed

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
+
+    #{$se23} & {
+      // stylelint-disable -- Polaris component custom properties
+      --pc-index-table-checkbox-offset-left: var(--p-space-3);
```
When tophatting polaris uplift, the cell content does not align with the cell table header also resulting in different spacing amount between the checkbox and table header vs. checkbox and table row content

https://github.com/Shopify/polaris/assets/15054990/ca5ec158-5ead-43ff-9432-9439b70a3555



<img width="629" alt="Screenshot 2023-09-25 at 3 58 08 PM" src="https://github.com/Shopify/polaris/assets/15054990/1f817de0-dcf8-4dcf-9c39-3f322911528e">

oksanashopify replied:
I left a comment in PR description, that "⚠️ Note: if rows looks broken - please, reload the page".
For some reason when toggle `polarisSummerEditions2023` that's what happens. And after we reload the page - everything looks fine. 
Maybe I will rebase from summer editions, as you recommended, to see if it would change the behaviour.
If this behaviour will still persist, I will clarify with Polaris team this question.

oksanashopify replied:
With @shopify/polaris@12.0.0 [has shipped](https://shopify.slack.com/archives/C4Y8N30KD/p1696832186682759) we don't need to duplicate styles and use that flag anymore 🎉

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -259,6 +259,48 @@ $loading-panel-height: 53px;
+        var(--pc-index-table-checkbox-width)
+    );
+
+    #{$se23} & {
```
Since summer editions will be cleaned up by the end of the week, would it make sense to rebase this branch off of the summer editions clean-up to reduce the amount of changes we are making and the number of cases we need to tophat as a result?

chloerice replied:
+1, I've updated the base branch of the PR. You'll need to rebase off of `v12.x.x` as some `IndexTable changes have been made in these files in v12` (like the removal of the IndexFilterManager).

oksanashopify replied:
Done! Thank you!

chloerice commented on polaris-react/src/components/IndexTable/components/Row/Row.tsx:
```diff
@@ -42,7 +42,7 @@ export interface RowProps {
   selectionRange?: Range;
   /**
-   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header.
+   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header. A "child" row represents the nested structure.
```
```suggestion
   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header. Rows of tyoe "child" row are indented.
```

oksanashopify replied:
Fixed.

chloerice commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -3929,3 +3933,1346 @@ export function WithSubHeaders() {
   );
 }
+
+export function WithNestedRows() {
```
To add this example to the website, you'll need to add a file for it in `polaris.shopify.com/pages/examples` (duplicate and existing file, rename it, and replace the code with this story's code).

oksanashopify replied:
Than you, @chloerice! Done.
Can I check it somehow in my PR how it looks?

chloerice replied:
Yes, you can run `yarn build && yarn turbo run dev --filter=polaris.shopify.com` and it will build the packages and site and then open localhost:3000 for you in a new tab/window when the build is successful.

chloerice commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -3929,3 +3933,1346 @@ export function WithSubHeaders() {
+
+  const resourceName = {
+    singular: 'customer',
+    plural: 'customers',
```
```suggestion
    singular: 'product',
    plural: 'products',
```

oksanashopify replied:
Done to all related stories

chloerice commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -3929,3 +3933,1346 @@ export function WithSubHeaders() {
+  const {selectedResources, allResourcesSelected, handleSelectionChange} =
+    useIndexResourceState(rows, {resourceFilter: ({disabled}) => !disabled});
+
+  const orders = groupRowsBy(
```
```suggestion
  const products = groupRowsBy(
```

oksanashopify replied:
renamed to `groupedProducts`

chloerice reviewed (commented):
Hey @oksanashopify This is looking good! Left some comments on the story, let's make sure it's added to the website examples, and update the code to use product/variants instead of order/customers. You'll need to rebase on `v12.x.x` to remove the unrelated changes from code that was removed in in the pending v12 release.

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -7,7 +7,7 @@ import type {
 import {
   Icon,
-  HorizontalStack,
+  InlineStack,
```
With @shopify/polaris@12.0.0 [has been shipped](https://shopify.slack.com/archives/C4Y8N30KD/p1696832186682759) we no longer have `HorizontalStack` component, so I replaced it with `InlineStack`.
Also removed unused imports

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -4002,3 +4002,1435 @@ export function WithSubHeaders() {
+        resourceName={resourceName}
+        itemCount={rows.length}
+        headings={columnHeadings as IndexTableProps['headings']}
+        lastColumnSticky
```
Seems like `lastColumnSticky` doesn't work with new Polaris version in all of the stories forIndexTable, same here.
Does anybody can confirm the same? Should I create an issue for that?

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -792,6 +819,22 @@ $loading-panel-height: 53px;
+        var(--pc-index-table-checkbox-width)
+    );
+    position: relative;
+    left: var(--pc-index-table-checkbox-child-offset);
```
I, personally, not happy how the indentation looks like for unselectable table with thumbnails
<img width="1493" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/ad39f98d-50b3-49f0-8eb9-d061defc4a72">

It is unclear how to calculate indentation for unselectable rows, which don't have checkboxes (`With Nested Rows Non Selectable`, `With Nested Rows With Thumbnails Non Selectable` stories). We also can't rely on thumbnail sizes.
In my example I used the same const ` --pc-index-table-checkbox-child-offset:` which I use for selectable tables just to be consistent. 

My other idea is - use half of ` --pc-index-table-checkbox-child-offset:` indentation and then align thumbnails if needed wrapping them in [Polaris Box](https://polaris.shopify.com/components/layout-and-structure/box) and adding padding on consumer level, similarly it was done for two last stories for Variants card sturcture

I would appreciate any ideas. Thank you!

oksanashopify commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+  };
+
+  const {selectedResources, allResourcesSelected, handleSelectionChange} =
+    useIndexResourceState(rows as unknown as {[key: string]: unknown}[], {
```
I used this type conversion just for the sake of this example do not give typescript errors, as https://github.com/Shopify/polaris/blob/93defb11bd8e4c5f120265a0d3c0aad8a35f7771/polaris-react/src/utilities/use-index-resource-state.ts#L25 uses `useIndexResourceState<T extends {[key: string]: unknown}`

kvnify reviewed (approved):
This seems inline with the use cases we'd considered. I don't have any blocking concerns. The mobile layout feels like it might be too subtle, but you are certainly limited there.

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -4002,3 +4002,1435 @@ export function WithSubHeaders() {
+          disabled={disabled}
+          accessibilityLabel={`Select all products wich has color ${color}`}
+        >
+          <IndexTable.Cell scope="col" id={subheaderId}>
```
Parent row should not have a design treatment of a subheader (no subdued background), at least in our use case, can we update it in storybook for tophatting?

<img width="1897" alt="Screenshot 2023-10-20 at 11 37 10 AM" src="https://github.com/Shopify/polaris/assets/15054990/f684e4e5-bff0-4925-b470-714cd6d01f1c">

oksanashopify replied:
I updated two last storybooks for tophating without subheader treatment. Thank you for the idea, @tatianau !

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -4002,3 +4002,1435 @@ export function WithSubHeaders() {
+  );
+}
+
+export function WithNestedRowsNonSelectable() {
```
This story (and all non-selectable stories) seems to have a broken background on the parent row, do you know if it's a regression caused by the implementation or is there just something wrong in the storybook component structure of that use case that we need to update?


https://github.com/Shopify/polaris/assets/15054990/af8def03-b1b2-4b8a-9531-ff22c7b98488

oksanashopify replied:
It was a regression caused by  @shopify/polaris@12.0.0 [has been shipped](https://shopify.slack.com/archives/C4Y8N30KD/p1696832186682759)
I shipped a fix for Polaris https://github.com/Shopify/polaris/pull/10960 and rebased this branch from the main, so this issue should not be gone now

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -217,6 +217,33 @@ $loading-panel-height: 53px;
+      var(--pc-index-table-checkbox-offset-left) +
+        var(--pc-index-table-checkbox-width)
+    );
+    --pc-index-table-shifted-checkbox: 32;
```
If this is a z-index, could we updater the custom property name to reflect that for readability?

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -802,6 +829,22 @@ $loading-panel-height: 53px;
     }
   }
 
+  &.Table-sticky .TableRow-child .TableCell:first-child > span {
```
What is this `span` selector targeting? Using tags as selectors is too generic and cause issues as it doesn't scale well, is there another way to target the element?

oksanashopify replied:
Good point! Thank you! I found another way

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -217,6 +217,33 @@ $loading-panel-height: 53px;
+        padding-left: var(--pc-index-table-checkbox-child-offset);
+      }
+    }
+  }
```
It seems like `padding-left` with this value has already been applied and removing it doesn't change anything. Why is it needed twice here? Is there another way?

```suggestion
    .TableCell-first + .TableCell {
      padding-left: var(--pc-index-table-checkbox-child-offset);
    }
  }
```

https://github.com/Shopify/polaris/assets/15054990/b3391c75-4286-42ef-989d-ea3c63434464

oksanashopify replied:
Thank you for pointing on that! Seems there is no need for specifying it twice!

tatianau commented on polaris-react/src/components/IndexTable/components/Row/Row.tsx:
```diff
@@ -32,7 +32,7 @@ export interface RowProps {
   selectionRange?: Range;
   /**
-   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header.
+   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header. Rows of type "child" row are indented.
```
```suggestion
   * Indicates the relationship or role of the row's contents. A "subheader" row displays the same as the table header. Rows of type "child" are indented.
```

tatianau reviewed (commented):
I was unable to tophat the [snapshot](https://github.com/Shopify/polaris/pull/10599#issuecomment-1773280100) in the admin. 

I'm not exactly sure why it errors with something unrelated to this PR, my hunch is that it could be because `oa-polaris-indextable-nested-child-rows` branch is based off of `main` which is already `v12` and running it in web errors since it's a major version upgrade with breaking changes that requires some fixing in the admin before we can even run our nested rows work.  So if that's true, we'll have to wait until `shopify/web` is upgraded to `v12` to be able to test the changes to the nested rows here. @Shopify/polaris-eng please correct me if I'm wrong.

In the meantime, I left some comments/questions.

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -217,6 +217,60 @@ $loading-panel-height: 53px;
+        left: var(--pc-index-table-checkbox-child-offset-sm);
+      }
+
+      &::before {
```
Fixes [bug with scrollable tables](https://github.com/Shopify/polaris/pull/10599#issuecomment-1771100548). Thank you @tatianau , for the idea!! ❤️

oksanashopify committed:
started to fix bg color for tables with row status

fixed bg for hovered rows with status

fixed active tone for nested rows

fixed selected hovered state for nested rows with status

fixing bg color for nested zebra rows

fixed hover and selected bg for zebra rows

fixed story with zebra styles

added columns to story with sticky column

oksanashopify commented on polaris-react/src/components/IndexTable/IndexTable.stories.tsx:
```diff
@@ -1835,6 +1948,7 @@ export function WithStickyLastColumn() {
         lastColumnSticky
       >
         {rowMarkup}
+        {nestedRowMarkup}
```
@chloerice suggested to add nested rows to existing stories, so each time developers introduce new changes  - they will be able to see how it works with nested rows. Please let me know if  I should create create separate stories instead.

oksanashopify committed:
fixed stories with sticky last column

tatianau commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+    return (
+      <Fragment key={subheaderId}>
+        <IndexTable.Row
+          rowType="subheader"
```
If this is an example of `nested rows` with `range selection` - it should not contain a `subheader` design treatment as we already have that in the [docs](https://polaris.shopify.com/components/tables/index-table?example=index-table-with-subheaders).

```suggestion
          rowType="parent"
```

If the `parent` type doesn't exist and/or there is no need for it to exist - we can default to `data` instead of `subheader`.

tatianau commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+  );
+
+  const rowMarkup = Object.keys(groupedProducts).map((color, index) => {
+    const {products, position, id: subheaderId} = groupedProducts[color];
```
This is not a subheader, let's call it a parentId
```suggestion
    const {products, position, id: parentId} = groupedProducts[color];
```

kyledurand replied:
Or `productId`?

oksanashopify replied:
@kyledurand can this example also be tested somehow the similar way I run stories?

kyledurand replied:
You you can build the site locally. Should just be `yarn dev --filter=polaris.shopify.com`

oksanashopify replied:
Thank you @kyledurand ! For some reasons I'm getting an error and site never opens even on polaris `main` branch. 🤔 
<img width="1079" alt="image" src="https://github.com/Shopify/polaris/assets/104942025/c3880f56-61bb-403c-b191-a111abf0f2f1">

oksanashopify replied:
resolved in  polaris channel

tatianau commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+        <IndexTable.Row
+          rowType="subheader"
+          selectionRange={rowRange}
+          id={`Subheader-${index}`}
```
Let's remove all the subheader-related things from the docs

```suggestion
          id={`Parent-${index}`}
```

tatianau commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+          position={position}
+          selected={selected}
+          disabled={disabled}
+          accessibilityLabel={`Select all products wich has color ${color}`}
```
```suggestion
          accessibilityLabel={`Select all products which have color ${color}`}
```

tatianau commented on polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx:
```diff
@@ -0,0 +1,233 @@
+        </IndexTable.Row>
+        {products.map(
+          ({id, size, quantity, price, position, disabled}, rowIndex) => {
+            return (
```
```suggestion
        {products.map(
          ({id, size, quantity, price, position, disabled}, rowIndex) => (
```

tatianau commented on polaris-react/src/components/IndexTable/IndexTable.scss:
```diff
@@ -783,6 +1078,23 @@ $loading-panel-height: 53px;
 }
 
 .Table-unselectable {
+  // for a consistency we calculate indentation for nested rows the same way as for selectable table
+  // stylelint-disable -- Polaris component custom properties
+  --pc-index-table-checkbox-width: var(--p-space-500);
+  // checkbox width is bigger for 2px for small devices
+  --pc-index-table-checkbox-width-sm: calc(
+    var(--pc-index-table-checkbox-width) + var(--p-width-050)
+  );
+  --pc-index-table-checkbox-padding-left: var(--p-space-300);
+  --pc-index-table-checkbox-child-offset: calc(
+    var(--pc-index-table-checkbox-width) +
+      var(--pc-index-table-checkbox-padding-left)
+  );
+  --pc-index-table-checkbox-child-offset-sm: calc(
+    var(--pc-index-table-checkbox-width-sm) +
+      var(--pc-index-table-checkbox-padding-left)
+  );
+  // stylelint-enable -- Polaris component custom properties
```
If this is the same as in the case of [selectable rows](https://github.com/Shopify/polaris/pull/10599/files?show-viewed-files=true&file-filters%5B%5D=#diff-e9d0c2a006c030f315115b92e46c3b44667bef0b5e909e1b3af852e6eb2b605bR269-R285) - can the custom properties and the final offset value calculation be defined in one place for all since it's being reused? 

For example, could they be defined at the [.Table](https://github.com/Shopify/polaris/pull/10599/files?show-viewed-files=true&file-filters%5B%5D=#diff-e9d0c2a006c030f315115b92e46c3b44667bef0b5e909e1b3af852e6eb2b605bL93) level and then used in `&.TableRow-child` and `.Table-unselectable`? 

If we can take [Frame](https://github.com/Shopify/polaris/blob/64ee750393a67ec271908e597403cc2035e35660/polaris-react/src/components/Frame/Frame.scss) as an example -  the `--pc-frame-button-size` custom property is defined at `.Frame` and then [used](https://github.com/Shopify/polaris/blob/64ee750393a67ec271908e597403cc2035e35660/polaris-react/src/components/Frame/Frame.scss#L95-L97) further on in `.NavigationDismiss`.

oksanashopify replied:
Not sure if it is possible to do without making changes to current selectors structure, as `.Table-unselectable` selector lives  on the same level  with `.Table` selector within `IndexTable.scss` 🤔

tatianau reviewed (approved):
The changes look great! 👏  I appreciate the inclusion of the spin instance allowing us to see the feature in action in the variants card. 

Left some non-blocking suggestions, and I'll defer to the Polaris team's expertise in providing the feedback on the code. It looks great from a consumer perspective.

kyledurand commented on .changeset/thin-ladybugs-film.md:
```diff
@@ -0,0 +1,5 @@
+'@shopify/polaris': patch
+---
+
+IndexTable nested rows
```
Added `child` type to `IndexTable` row

kyledurand reviewed (approved):
The code looks good so I'm going to approve to unblock you since this has been open for a while and can be iterated on.

One thing I'd like to get some feedback from @Shopify/polaris-ux is on small screens where the sticky cells are offset causing a snaky type of border thing

![image](https://github.com/Shopify/polaris/assets/6844391/402b7a07-ed8f-496e-a530-9eb3053c5d27)

oksanashopify committed:
added new example to indextable page and small renamong after reviews

oksanashopify committed:
example code cleanup

oksanashopify committed:
updated documentation
